### PR TITLE
@broskoski & @jtotoole: Huge refactor of Artsy Passport #WIP

### PR DESCRIPTION
**Pull request by craigspaeth**
This breaks AP up into a bunch of more sensible modules instead of the big monolith that it is today. In this journey I also implemented...

* Fixes for various "lifecycle" logic
  * Edge cases around signing up with twitter, then trying to link from a different account
  * Avoid the awkward double-accept from Twitter on sign up
* Cover more auth use cases and push more related logic in here
  * Linking accounts
  * Redirecting back to the page you logged in from
* Beginning of implementing login/signup via LinkedIn

To refresh your memories on what Artsy Passport does at a granular level—it is an express app that gets mounted into Force & MG. It provides all of the auth logic up to the actual UI of login/etc. forms—leaving Force & MG to come up with their own templates/styling and AP handling the rest. The AP express app...
* Has various "Artsy auth routing business logic" like ensuring a user who signed up through Twitter is redirected to the "one last step" page to enter their email, or allowing the mobile app to log in by passing their Artsy access token through headers.
* Combines API calls to Artsy with social provider OAuth dances powered by Passport.js to power login/sign up/linking
* Uses the Artsy API to fetch user data, cache it in the session, and add locals like `req.user` and `sd.CSRF_TOKEN`
* Does many more surprisingly numerous login/sign up/social linking things I'm probably forgetting

This is still WIP b/c I need to rehaul the tests, but thought I would put it up here to gaze upon. This shit is confusing and complex, so I don't expect a full code review. Do your best and forget the rest.

### CODE REVIEW

File: lib/app/index.coffee [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+app = express()
+opts = require '../options'
+twitterLastStep = require './twitter_last_step'
+{ onLocalLogin, onLocalSignup, beforeSocialAuth,
+  afterSocialAuth, ensureLoggedInOnAfterSignupPage } = require './lifecycle'
+{ denyBadLogoutLinks, logout } = require './logout'
+{ headerLogin, trustTokenLogin } = require './token_login'
+addLocals = require './locals'
+
+module.exports = ->
```
Comment (broskoski) [inline lib/app/index.coffee]: Nice, this is so much more organized

File: lib/app/lifecycle.coffee [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    else if req.xhr and req.user?
+      res.send { success: true, user: req.user.toJSON() }
+    else if req.xhr and not req.user?
+      res.send 500, { success: false, error: "Missing user." }
+    else if err
+      next err
+    else
+      redirectBack req, res
+
+@onLocalSignup = (req, res, next) ->
```
Comment (broskoski) [inline lib/app/lifecycle.coffee]: And way better naming :+1:

File: lib/app/redirectback.coffee [hunk 3]
```diff
@@ -0,0 +1,45 @@
+#
+# Redirects back based on query params, session, or w/e else.
+# Code stolen from Force, thanks @dzucconi!
```
Comment (broskoski) [inline lib/app/redirectback.coffee]: :wave

File: lib/app/twitter_last_step.coffee [hunk 4]
```diff
@@ -0,0 +1,52 @@
+#
+# Twitter last step logic. Used to ensure we have an email from users that
+# sign up with Twitter.
```
Comment (broskoski) [inline lib/app/twitter_last_step.coffee]: Yeah, def better to break this guy out into its own file.

File: lib/index.coffee [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+opts = require './options'
+setupApp = require './app'
+setupPassport = require './passport'
+artsyXapp = require 'artsy-xapp'
+
+module.exports = (options) =>
+  _.extend opts, options
+  setupPassport()
+  setupApp()
+
```
Comment (broskoski) [inline lib/index.coffee]: :ok_hand:

**Comment by broskoski** [comment]
Much needed overhaul, looks much more manageable to me! :+1: :balloon: A+

File: lib/app/logout.coffee [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+  accessToken = req.user?.get('accessToken')
+  req.logout()
+  request
+    .del("#{opts.ARTSY_URL}/api/v1/access_token")
+    .send(access_token: accessToken)
+    .end (error, response) ->
+      if req.xhr
+        res.status(200).send msg: 'success'
+      else
+        res.redirect req.query['redirect-to'] or '/'
```
Comment (craigspaeth) [inline lib/app/logout.coffee]: Note to self: Use `redirectBack` here.

File: lib/app/logout.coffee [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+    next()
+  else
+    next new Error "Malicious logout link."
+
+@logout = (req, res, next) ->
+  accessToken = req.user?.get('accessToken')
+  req.logout()
+  request
+    .del("#{opts.ARTSY_URL}/api/v1/access_token")
+    .send(access_token: accessToken)
```
Comment (craigspaeth) [inline lib/app/logout.coffee]: Also, note to self: use `set` here

File: lib/passport/callbacks.coffee [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+  if not err
+    done null, new opts.CurrentUser accessToken: res.body.access_token
+  # If there's no user linked to this account, create the user via the POST
+  # /user API. Then attempt to fetch the access token again from Gravity and
+  # recur back into this onAcccessToken callback.
+  else if msg.match('no account linked')?
+    req.artsyPassportSignedUp = true
+    request
+      .post(opts.ARTSY_URL + '/api/v1/user')
+      .send(_.extend params)
```
Comment (craigspaeth) [inline lib/passport/callbacks.coffee]: Unnecessary extend here

File: lib/passport/index.coffee [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+  passport.use new LinkedInStrategy
+    consumerKey: opts.LINKEDIN_KEY
+    consumerSecret: opts.LINKEDIN_SECRET
+    passReqToCallback: true
+    callbackURL: "#{opts.APP_URL}#{opts.linkedinCallbackPath}"
+    profileFields: [
+      'id', 'first-name', 'last-name', 'email-address', 'headline', 'location',
+      'industry', 'summary', 'specialties', 'positions', 'public-profile-url'
+    ]
+  , callbacks.linkedin
```
Comment (dzucconi) [inline lib/passport/index.coffee]: I think these things will error if the ENV variables are missing? I remember having to include social id/secrets in Vanity just to get it to stop yelling.

Comment (craigspaeth) [inline lib/passport/index.coffee]: Good call! I can conditionally mount strategies based on config existential checks. Thanks for pointing that out.

File: lib/app/lifecycle.coffee [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+  request
+    .post(opts.ARTSY_URL + '/api/v1/user')
+    .set('X-Xapp-Token': artsyXapp.token)
+    .send(
+      name: req.body.name
+      email: req.body.email
+      password: req.body.password
+    ).end (err, res) ->
+    errMsg = if res.status isnt 201 then res.body.message else err?.text
```
Comment (craigspaeth) [inline lib/app/lifecycle.coffee]: Indentation is wrong!

**Comment by jtotoole** [comment]
@craigspaeth Looks great! You've earned yourself an endorsement or two

**Comment by craigspaeth** [comment]
Finally wrapped up the refactor with a decent set of integration and unit tests—still plenty of pending unit tests to further fill in, but I think that wraps up my epic journey. I will try to look into integrating this back in to our apps now and unblock James from linked in integration.

**Comment by jtotoole** [comment]
:clap: :clap: :clap:

**Pull request merged**